org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResource.java

```diff
@@ -0,0 +1,136 @@
+/**
+ * Copyright (c) 2014, 2020 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.resource.persistence;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.log4j.Logger;
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.xtext.linking.lazy.LazyLinkingResource;
+import org.eclipse.xtext.resource.IResourceDescription;
+import org.eclipse.xtext.util.internal.Stopwatches;
+
+import com.google.inject.Inject;
+
+/**
+ * A resource implementation that can load itself from ResourceStorage.
+ * 
+ * @author Sven Efftinge - Initial contribution and API
+ */
+public class StorageAwareResource extends LazyLinkingResource {
+	public static final String UNRESOLVABLE_FRAGMENT = "UNRESOLVABLE";
+
+	private static final Logger LOG = Logger.getLogger(StorageAwareResource.class);
+
+	@Inject(optional = true)
+	private IResourceStorageFacade resourceStorageFacade;
+
+	@Inject
+	private PortableURIs portableURIs;
+
+	private boolean isLoadedFromStorage = false;
+
+	private IResourceDescription resourceDescription = null;
+
+	@Override
+	public void load(Map<?, ?> options) throws IOException {
+		if (!isLoaded && !isLoading && resourceStorageFacade != null
+				&& resourceStorageFacade.shouldLoadFromStorage(this)) {
+			if (StorageAwareResource.LOG.isDebugEnabled()) {
+				LOG.debug("Loading " + getURI() + " from storage.");
+			}
+			try {
+				ResourceStorageLoadable in = resourceStorageFacade.getOrCreateResourceStorageLoadable(this);
+				loadFromStorage(in);
+				return;
+			} catch (IOException e) {
+				if (contents != null) {
+					contents.clear();
+				}
+				if (eAdapters != null) {
+					eAdapters.clear();
+				}
+				unload();
+			}
+		}
+		super.load(options);
+	}
+
+	public void loadFromStorage(ResourceStorageLoadable storageInputStream) throws IOException {
+		if (storageInputStream == null) {
+			throw new NullPointerException("storageInputStream");
```

Comment:
You might like to change the detailed message to `the parameter storageInputStream must not be null` or something better.